File: Android/Blink/libandroid/src/main/java/net/qiujuer/blink/core/ReceiveParser.java
package net.qiujuer.blink.core;

import net.qiujuer.blink.box.ByteReceivePacket;
import net.qiujuer.blink.box.FileReceivePacket;
import net.qiujuer.blink.box.StringReceivePacket;

import java.io.File;

/**
 * Blink receive entity parse
 */
public class ReceiveParser {
    private long mId = 0;
    protected Resource mResource;

    public ReceiveParser(Resource resource) {
        mResource = resource;
    }

    public ReceivePacket<?> parseReceive(int type, int len) {
        final long id = ++mId;
        ReceivePacket<?> entity = null;
        switch (type) {
            case BlinkPacket.Type.STRING:
                entity = new StringReceivePacket(id, type, len);
                break;
            case BlinkPacket.Type.BYTES:
                entity = new ByteReceivePacket(id, type, len);
                break;
            case BlinkPacket.Type.FILE:
                File file = mResource.create(id);
                if (file != null)
                    entity = new FileReceivePacket(id, type, len, file);
                break;
        }
        return entity;
    }
}

File: C Sharp/Blink/Blink/Core/ReceiveParser.cs
using Net.Qiujuer.Blink.Box;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Net.Qiujuer.Blink.Core
{
    /// <summary>
    /// Blink receive packet parse
    /// </summary>
    public class ReceiveParser
    {
        private long mId = 0;
        protected Resource mResource;

        public ReceiveParser(Resource resource)
        {
            mResource = resource;
        }

        public ReceivePacket ParseReceive(int type, int len)
        {
            long id = ++mId;
            ReceivePacket packet = null;
            switch (type)
            {
                case Type.STRING:
                    packet = new StringReceivePacket(id, type, len);
                    break;
                case Type.BYTES:
                    packet = new ByteReceivePacket(id, type, len); ;
                    break;
                case Type.FILE:
                    String file = mResource.Create(id);
                    if (file != null)
                        packet = new FileReceivePacket(id, type, len, file);
                    break;
            }
            return packet;
        }

        public static class Type
        {
            public const int STRING = 0;
            public const int BYTES = 1;
            public const int FILE = 2;
        }
    }
}
